### PR TITLE
feat(async): add support for async dependencies

### PR DESCRIPTION
**Pull request by patrickhulce**
# Usage
## Define an Async Module

```
exports = module.exports = function createDbConnetion(conf, mysql) {
  return mysql.connect(conf.sqlConnectionString).then(...);
};

exports['@async'] = true; // async because the module it returns is a within a promise
exports['@require'] = ['conf', 'mysql'];
```
## Use an Async Module

```
IoC.createAsync('database').then(function (db) {
  db.execute('SELECT * FROM my_table');
}):
```
## Depend on an Async Module

```
exports = module.exports = function usersModel(db) {
  return {
    create: function (name, email) {
      return db.execute('INSERT INTO users ...');
    }
  };
};

exports['@async'] = true; // async because it depends on an async module
exports['@require'] = ['database'];
```

### CODE REVIEW

**Comment by shaharke** [comment]
Nice! 👍

File: lib/container.js [hunk 1]
```diff
@@ -133,10 +141,21 @@ Container.prototype.create = function(id, parent) {
...
+    return Promise.resolve(spec.create(this)).then(function (obj) {
+      this.emit('create', obj, spec);
+      return obj;
+    }.bind(this));
+  } else {
+    if (spec.isAsync) {
+      throw new Error('Container#create cannot be called on async dependency: "' + id + '"');
+    }
+
   var obj = spec.create(this);
```
Comment (shaharke) [inline lib/container.js]: Indentation issue or is it just Github?

Comment (patrickhulce) [inline lib/container.js]: whoops, yeah that's indentation issue. thanks!

**Comment by xduseko** [comment]
@jaredhanson, is this going to be merged? released?

**Comment by patrickhulce** [comment]
@jaredhanson any feedback and/or thoughts on this, positive or negative?

**Comment by jaredhanson** [comment]
@patrickhulce I pulled your changes into a local branch, `promises` here: https://github.com/jaredhanson/electrolyte/tree/promises

The only thing I wasn't to fond of, was that the async and sync paths were different, and it wasn't clear at all as a developer using the IoC container, if a dependency was going to be async or not.  For example, lets say you had a component that was sync today, and then tomorrow took an async dependency.  When updating, you could no longer `IoC.create` that dependency, and had to switch to `IoC.createAsync`.

I do like the ability to create components asynchronously.  It seems like the best approach is to simply make that the default.  That's what I've done in the branch above.  It is stable, I just need to review test coverage.

If you want to pull my branch, look it over, and then update any tests and documentation, I would really appreciate it.  This will be released as `electrolyte@0.4.0`.

Thanks for your PR!

**Comment by FWoelffel** [comment]
Is this still relevant?

**Comment by patrickhulce** [comment]
@FWoelffel no the core functionality has since been merged to master 🎉  I'll close.

**Pull request closed without merging**